### PR TITLE
Autocomplete should support entities without bundle key.

### PR DESCRIPTION
**Pull request by RoySegall**
#351

I'll start working on a test for that.

### CODE REVIEW

**Comment by RoySegall** [comment]
@amitaibu This is ready for review.

File: restful.module [hunk 1]
```diff
@@ -197,6 +197,16 @@ function restful_menu() {
 }
 
 /**
+ * Implements hook_entity_info_alter().
+ *
+ * Define the name property as the label of the user. We need this inorder for
+ * users endpoint to be able to autocomplete users.
+ */
+function restful_entity_info_alter(&$entity_info) {
```
Comment (amitaibu) [inline restful.module]: I don't think we would like to add that - it's too general. I think it's better to override the autocomplete function in the user resource, and making sure it queries by the `name` property.

**Comment by RoySegall** [comment]
@amitaibu Can you have a second look?

File: plugins/restful/RestfulDataProviderEFQ.php [hunk 2]
```diff
@@ -47,6 +47,18 @@ public function getEntityType() {
...
   /**
+   * Get the entity info for the current entity the endpoint handling.
+   *
+   * @param null $type
+   *   The entity type. Optional.
+   * @return array
+   *   The entity info.
+   */
+  public function getEntityInfo($type = NULL) {
+    return entity_get_info($type ? $type : $this->getEntityType());
```
Comment (e0ipso) [inline plugins/restful/RestfulDataProviderEFQ.php:58]:
``` php
return entity_get_info($type ?: $this->getEntityType());
```

**Comment by e0ipso** [comment]
I'm not sure I understand the problem. Will other entities (that don't have bundles and don't have resources shipped by RESTful core) suffer from the same problem? Will they have to override getEntityInfo?

If so, we should find a way to avoid doing that since users won't know about it and will file bug reports for that. Is that possible?

File: plugins/restful/RestfulDataProviderEFQ.php (hunk 2 above)
Comment (RoySegall) [inline plugins/restful/RestfulDataProviderEFQ.php:58]: `return entity_get_info($type ?: $this->getEntityType());` won't work in php 5.3

**Comment by RoySegall** [comment]
Entities without bundles are vocabulary and user. User is the only entity that don't define the label key and there are three ways i can think that will solved this:
1. Patch Drupal 7 but i don't think any one will merge this one in.
2. Use hook_entity_info_alter and just fix the user entity as i deed at the beginning
3. Use the current solution - the entity get info method. This look kind of a hack but in the long term if some one want to change the label key for the automplete he can do this in the controller.

File: plugins/restful/RestfulDataProviderEFQ.php (hunk 2 above)
Comment (e0ipso) [inline plugins/restful/RestfulDataProviderEFQ.php:58]: Oh! I thought it was available for php5.3

http://php.net/manual/en/language.operators.comparison.php#114872

**Comment by e0ipso** [comment]
> Entities without bundles are vocabulary and user

There can be custom defined entities that don't have bundles, right?

> User is the only entity that don't define the label key

Is there a solution to fix this problem for all the entities that don't define the label key instead of only fixing it for the user entity?

**Comment by RoySegall** [comment]
> There can be custom defined entities that don't have bundles, right?

Sure. For non fieldable entities a bundle is not mandatory.

> Is there a solution to fix this problem for all the entities that don't define the label key instead of only fixing it for the user entity?

I can't think on generic way to solve this. For example message don't define a label for the message entity. I can't see an autocomplete for messages but this is one example for the problem that maybe controllers would like to override the entity info.

**Comment by e0ipso** [comment]
Sorry it took so long to merge. It's merged now.

Thanks!!!!

**Pull request closed without merging**